File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> ranh = new HashMap<>();
        HashMap<Character, Integer> magh = new HashMap<>();  
        
        for(int i = 0; i < ransomNote.length(); i++) {
            ranh.merge(ransomNote.charAt(i), 1, Integer::sum);
        }
        for(int i = 0; i < magazine.length(); i++) {
            magh.merge(magazine.charAt(i), 1, Integer::sum);
        }
        
        for(int i = 0; i < ransomNote.length(); i++) {
            if(!magh.containsKey(ransomNote.charAt(i)))          {
               return false; 
            }
            if(!(magh.get(ransomNote.charAt(i)) >= ranh.get(ransomNote.charAt(i)))) {
                return false;
            }
        }
        
        return true;
    }
}

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        rmap = {}
        mmap = {}

        for r in ransomNote:
            if r in rmap:
                rmap[r] += 1
            else:
                rmap[r] = 1
        for m in magazine:
            if m in mmap:
                mmap[m] += 1
            else:
                mmap[m] = 1
        
        for i in ransomNote:
            if i not in mmap:
                return False
            if rmap[i] <= mmap[i]:
                continue
            else:
                return False
        return True